torchvision/csrc/cpu/ROIAlign_cpu.cpp

```diff
@@ -134,18 +135,23 @@ void ROIAlignForward(
     int roi_batch_ind = offset_rois[0];
 
     // Do not using rounding; this implementation detail is critical
-    T roi_start_w = offset_rois[1] * spatial_scale;
-    T roi_start_h = offset_rois[2] * spatial_scale;
-    T roi_end_w = offset_rois[3] * spatial_scale;
-    T roi_end_h = offset_rois[4] * spatial_scale;
-    // T roi_start_w = round(offset_rois[0] * spatial_scale);
-    // T roi_start_h = round(offset_rois[1] * spatial_scale);
-    // T roi_end_w = round(offset_rois[2] * spatial_scale);
-    // T roi_end_h = round(offset_rois[3] * spatial_scale);
+    T offset = aligned ? (T)0.5 : (T)0.0;
+    T roi_start_w = offset_rois[1] * spatial_scale - offset;
+    T roi_start_h = offset_rois[2] * spatial_scale - offset;
+    T roi_end_w = offset_rois[3] * spatial_scale - offset;
+    T roi_end_h = offset_rois[4] * spatial_scale - offset;
 
     // Force malformed ROIs to be 1x1
     T roi_width = std::max(roi_end_w - roi_start_w, (T)1.);
     T roi_height = std::max(roi_end_h - roi_start_h, (T)1.);
+    if (aligned) {
+      AT_ASSERTM(
+          roi_width >= 0 && roi_height >= 0,
+          "ROIs in ROIAlign do not have non-negative size!");
+    } else { // for backward-compatibility only
+      roi_width = std::max(roi_width, (T)1.);
+      roi_height = std::max(roi_height, (T)1.);
+    }
     T bin_size_h = static_cast<T>(roi_height) / static_cast<T>(pooled_height);
     T bin_size_w = static_cast<T>(roi_width) / static_cast<T>(pooled_width);
 
```

Comment:
the version in detectron2 also has
```
count = std::max(roi_bin_grid_h * roi_bin_grid_w, 1.);
```
below at line 166, which is needed because now the bin can be empty and can lead to NaN outptus.
Same for the cuda version.

Reply:
@ppwwyyxx 
Sorry for miss.

And I have a question about ` T roi_width = std::max(roi_end_w - roi_start_w, (T)1.);`

In detecton2, Only RoIAlign cpu version handling negative size.
The cuda version just clamping `1.0`(using max)

Which do you think is the better way to raise an error or use max?

Reply:
Fix in https://github.com/pytorch/vision/pull/1908/commits/a7c2be6d0a3e22dcfe55ab715fcda91d57935175

Reply:
This change is not needed in backward because backward does not divide anything by `count`.